packages/@aws-cdk/cdk/lib/core/construct.ts

```diff
@@ -15,6 +15,12 @@ export interface IConstruct extends IDependable {
    * The construct node in the scope tree.
    */
   readonly node: ConstructNode;
+
```

Comment:
This should be defined on a dedicated interface. Definitely not on `IConstruct` and `Construct`. A construct is an abstract cloud component, a physical name is something that only low-level resources have.

Reply:
This is the crux of the philosophical discussion I alluded to in the PR description. Here's why I've defined it in `IConstruct`.

If you look in the PR, you will find the following code in the Pipeline Construct:

```ts
      // make sure the Bucket has a physical name if the Construct itself is managing it
      if (this.managedBucket) {
        this.artifactBucket.assignPhysicalNameIfNotSet();
      }
```

The reason that line is there is because I want the experience of using a cross-account CodePipeline to be as close to a regular, non-cross-account Pipeline as possible. When using regular CodePipeline, we create an S3 Bucket for the customer implicitly if they don't provide one:

```ts
pipeline = new codepipeline.Pipeline(this, 'Pipeline');
// an S3 Bucket is created behind the scenes
```

Of course, the created S3 Bucket does not have a physical name. Now, if you use that Pipeline in a cross-account manner, the code will explode saying that resource `Pipeline/ArtifactBucket/Resource` does not have a physical name set. I think that's a _terrible_ customer experience - showing an error that something we defined for the customer behind the scenes is invalid. We try to abstract things away, but suddenly we make them worry about it. If a customer doesn't know they can pass a Bucket to the Pipeline Construct, they will not know how to solve this problem at all!

This thing happens all over the place. For example, when creating a CodeBuild Project, we implicitly create a Role. When the Project is used in a cross-account matter, that Role needs to be given a physical name. Which means the customer needs to write something like this:

```ts
const projecRole = new iam.Role(this, 'ProjectRole',
  roleName: 'PhysicalName',
  assumedBy: new iam.ServicePrincipal('codebuild.amazonaws.com'),
});
```

I really hate that `assumedBy: new iam.ServicePrincipal('codebuild.amazonaws.com')` line. I don't want to force every customer that wants to have cross-account Pipelines to deal with this level of mundane details.

My solution is to introduce the 'managed by' concept - the `managedBucket ` property you see above the in the Pipeline Construct code. Meaning, if we are the ones that created the resource implicitly, we will assign it a physical name when needed, without bothering the customer about it. However, if they passed a Bucket explicitly, then we don't mess with it - the onus is on them to assign a physical name when required.

Does this make sense?

Reply:
But we could still have the infrastructure and the feature without defining it on `Construct`/`IConstruct`, no?

If we had an `AmazonPhysicalName` class, we could do:

```ts
class Bucket {
  private physicalName: AmazonPhysicalName;

  constructor(...) {
    this.physicalName = new AmazonPhysicalName(this, props.bucketName, { autogenerate: true }):

   this.bucketName = this.physicalName.physicalName;
   this.bucketArn = this.physicalName.physicalArn;
  }
}
```

Would this not have the same effect, but localized?

Reply:
And how would the Pipeline Construct assign the physical name to the Bucket in case it was a cross-account Pipeline in that scheme?

Reply:
Some code elided, obviously, but same way you're doing it now:

```ts
class Bucket {
  private physicalName: AmazonPhysicalName;

  constructor(...) {
    this.physicalName = new AmazonPhysicalName(this, props.bucketName, { autogenerate: true }):

   new CfnBucket(this, 'Resource', {
     bucketName: this.physicalName.physicalName,
   });

   this.bucketName = this.physicalName.physicalName;
   this.bucketArn = this.physicalName.physicalArn;
  }
}
```

Reply:
I still don't understand. What's the method the Pipeline Construct calls on Bucket, and in which class/interface does it live?

Reply:
I was thinking it would happen automatically if the `physicalName` token was attempted to be resolved and `autoresolve` was true.

Of course, it should happen during the `prepare()` phase, not the `resolve()` phase, which might make things a little more complicated. But if you inherit from `CloudFormationReference` `consumeFromStack()` will be called during `prepare()`, you could use that.

I'll agree it's a little nonobvious.

Reply:
Of course you could also just forward an `assignPhysicalName()` call:

```ts
class Bucket {
  public assignPHysicalName() {
    this.physicalName.assignIfPossible();
  }
}
```

Reply:
Moved to `Resource`.